src/libopensc/card-nqApplet.c

```diff
@@ -0,0 +1,556 @@
+	sc_format_apdu_ex(&apdu, 0x80, 0x2A, 0x9E, 0x9A, data, cb_data, out, outlen);
+	apdu.le = 256;
+	apdu.flags |= SC_APDU_FLAGS_CHAINING;
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+
+	if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00)
+	{
+		rv = apdu.resplen;
+	}
+	else if (apdu.sw1 == 0x61)
+	{
+		rv = apdu.sw2 == 0 ? 256 : apdu.sw2;
+	}
+	else
+	{
+		rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	}
+
+	LOG_FUNC_RETURN(card->ctx, rv);
+}
+
+static int nqapplet_check_sw(struct sc_card *card,unsigned int sw1,unsigned int sw2)
+{
+	const int nqapplet_error_count = sizeof(nqapplet_errors)/sizeof(struct sc_card_error);
+	int i;
+
+	LOG_FUNC_CALLED(card->ctx);
+	sc_log(card->ctx, "Checking sw1 = 0x%02x, sw2 = 0x%02x\n", sw1, sw2);
+  
+	for (i = 0; i < nqapplet_error_count; i++)
+	{
+		if (nqapplet_errors[i].SWs == ((sw1 << 8) | sw2))
+		{
+			LOG_TEST_RET(card->ctx, nqapplet_errors[i].errorno, nqapplet_errors[i].errorstr);
+		}
+	}
+
+	return iso_operations->check_sw(card, sw1, sw2);
+}
+
+static int nqapplet_get_data(struct sc_card *card, unsigned int id, u8 *resp, size_t cb_resp)
+{
+	struct sc_apdu apdu;
+	int rv;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	sc_format_apdu_ex(&apdu, 0x80, 0xB0, 0x00, (u8)id, NULL, 0, resp, cb_resp);
+	apdu.le = 256;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, rv, "APDU transmit failed");
+
+	if (apdu.sw1 == 0x90 && apdu.sw2 == 0x00)
+	{
+		rv = apdu.resplen;
+	}
+	else if (apdu.sw1 == 0x61)
+	{
+		rv = apdu.sw2 == 0 ? 256 : apdu.sw2;
+	}
+	else if (apdu.sw1 == 0x62 && apdu.sw2 == 0x82)
+	{
+		rv = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	}
+
+	LOG_FUNC_RETURN(card->ctx, rv);
+}
+
+static int nqapplet_select_file(struct sc_card *card, const struct sc_path *in_path, struct sc_file **file_out)
+{
+	LOG_FUNC_CALLED(card->ctx);
+
+	/* the applet does not support SELECT EF/DF except for SELECT APPLET.
+	In order to enable opensc-explorer add support for virtually selecting MF only */
+	if (in_path->type == SC_PATH_TYPE_PATH && in_path->len == 2 && memcmp(in_path->value, "\x3F\x00", 2) == 0 ) {
+		if ( file_out != NULL )   {
+			struct sc_file *file = sc_file_new();
+			if (file == NULL)
+			{
+				LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+			}
+			file->path = *in_path;
+			*file_out = file;
+			LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+		}
+	}
+	//TODO allow selecting Applet AID
+
+	LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+}
+
+static int nqapplet_card_ctl(sc_card_t* card, unsigned long cmd, void* ptr)
+{
+	switch (cmd)
+	{
+	case SC_CARDCTL_GET_SERIALNR:
+		if (card->serialnr.len) {
+			sc_serial_number_t* serial = (sc_serial_number_t*)ptr;
+			memcpy(serial->value, card->serialnr.value, card->serialnr.len);
+			serial->len = card->serialnr.len;
+			return SC_SUCCESS;
+		}
+		break;
+	}
+	return SC_ERROR_NOT_SUPPORTED;
+}
+
+
+struct sc_card_driver * sc_get_nqApplet_driver(void)
+{
+	sc_card_driver_t * iso_driver = sc_get_iso7816_driver();
```

Comment:
some of the functionality above looks almost like implemented in the iso7816 driver (e.g. get response, get data, get challenge). do you see some limitations in the iso driver that could be improved?

Reply:
I tried to use the iso7816 driver where it was possible, in fact I implemented only those functions which failed during testing. For instance the difference between iso7816_get_response and my nqapplet_get_response is the "cla" value. Unfortunately the applet uses 3 different cla values for its APDUs, so I could not rely on setting the card->cla.
This applet is not even trying to be iso compatible, so all of these incompatibilities are caused by the applet implementation and not the iso driver.